### ISSUE TITLE
Memory corruption

### ISSUE THREAD

**Issue by gnachman** [tmux/tmux#1870]
### Issue description

I was testing moving and linking windows in control mode and crashed the tmux server. I was able to reproduce a few times.

I tested at commit 26f274011096b9eacfd2753d8f7afcb63e796ab9, which is the current head of `3.0-rc`. Let me know if I should be trying on master or 3.0 instead.

The steps to reproduce are to create a few sessions and send commands like `move-window -s 3:@3 -t 5+` in quick succession to move and link windows between sessions. I was able to cause it to crash while logging. FWIW it was much easier to reproduce when not producing a log.

[tmux-server-42639.log](https://github.com/tmux/tmux/files/3504465/tmux-server-42639.log)

I was able to get a core dump:

```
* thread #1, stop reason = signal SIGSTOP
  * frame #0: 0x000000010001a533 tmux`control_notify_session_window_changed(s=0x000000010041e9f0) at control-notify.c:256:16 [opt]
    frame #1: 0x0000000100034ed5 tmux`notify_callback(item=0x000000010042a480, data=0x000000010042a420) at notify.c:128:3 [opt]
    frame #2: 0x0000000100011ec0 tmux`cmdq_next [inlined] cmdq_fire_callback(item=<unavailable>) at cmd-queue.c:371:10 [opt]
    frame #3: 0x0000000100011eaf tmux`cmdq_next(c=<unavailable>) at cmd-queue.c:449 [opt]
    frame #4: 0x0000000100046547 tmux`server_loop at server.c:229:11 [opt]
    frame #5: 0x0000000100039458 tmux`proc_loop(tp=0x00000001004002c0, loopcb=(tmux`server_loop at server.c:224)) at proc.c:200:42 [opt]
    frame #6: 0x0000000100045e19 tmux`server_start(client=<unavailable>, base=<unavailable>, lockfd=6, lockfile="/private/tmp/tmux-501/default.lock") at server.c:213:2 [opt]
    frame #7: 0x00000001000031a9 tmux`client_main [inlined] client_connect(base=<unavailable>, path="/private/tmp/tmux-501/default") at client.c:158:8 [opt]
    frame #8: 0x0000000100002f2f tmux`client_main(base=0x00000001003005f0, argc=0, argv=0x00007ffeefbff4c0, flags=<unavailable>) at client.c:266 [opt]
    frame #9: 0x000000010004f093 tmux`main(argc=0, argv=0x00007ffeefbff4c0) at tmux.c:374:7 [opt]
    frame #10: 0x00007fff60de03d5 libdyld.dylib`start + 1
    frame #11: 0x00007fff60de03d5 libdyld.dylib`start + 1
(lldb) f 0
frame #0: 0x000000010001a533 tmux`control_notify_session_window_changed(s=0x000000010041e9f0) at control-notify.c:256:16 [opt]
   253 				continue;
   254
   255 			control_write(c, "%%session-window-changed $%u @%u", s->id,
-> 256 			    s->curw->window->id);
   257 		}
   258 	}
(lldb) p s->curw
(winlink *) $0 = 0x0000000000000000
```

My `.tmux.conf` contains only commented-out lines:

```
#set-window-option -g automatic-rename off
#set-option -g set-titles on
```

### Required information

Please provide the following information:

* tmux version (`tmux -V`). tmux 3.0-rc4
* Platform (`uname -sp`). Darwin i386
* $TERM inside and outside of tmux (`echo $TERM`). xterm-256color
* Logs from tmux (`tmux kill-server; tmux -vv new`). See above.

**Comment by nicm** [comment]
Hmm. I can't reproduce, do you have a script or anything that will reproduce? It might help to do something like this so it also crashes without control mode:
~~~~
--- control-notify.c    10 Jul 2019 11:20:10 -0000      1.24
+++ control-notify.c    15 Aug 2019 08:32:14 -0000
@@ -248,6 +248,8 @@ control_notify_session_window_changed(st
 {
        struct client   *c;

+       if (s->curw == NULL || s->curw->window == NULL)
+               fatalx("NULL window");
        TAILQ_FOREACH(c, &clients, entry) {
                if (!CONTROL_SHOULD_NOTIFY_CLIENT(c))
                        continue;
~~~~
 The notification code only adds a reference to the session so there is nothing in theory stopping its current window going away, but I don't see how it can happen and it doesn't happen in the obvious cases or by copying the commands from your log file.

**Comment by gnachman** [comment]
I'm away from the machine I was using last night, and I can't get it to happen on this one. I did reproduce it three times last night, though. I think it must be a race condition of some sort because it happened only after a lot of repeatedly moving windows from one session to another in quick succession. I was sending multiple move-window and link-window commands on one line, like:
```
move-window -s "4:@6" -t "5:+"; move-window -s "4:@1" -t "5:+"
```

**Comment by nicm** [comment]
Hmm I'm a bit torn on this, it would be enough to add a check to control_notify_session_window_changed to stop it crashing but I suspect there is a deeper problem but without being able to reproduce I can't really tell what it is...

**Comment by AdrienLemaire** [comment]
Getting tmux crashes as well. Seems to occur more frequently when tmux panes are small, and when installing large softwares like chromium.

Running tmux -vvvv didn't output anything interesting in the log files.

dmesg:
```
[13508.470861] tmux: server[441303]: segfault at 0 ip 000055b11f3f6a79 sp 00007ffd1d725460 error 6 in tmux[55b11f3bb000+66000]
[13508.470870] Code: 18 48 c7 85 30 01 00 00 00 00 00 00 89 45 00 8b 43 1c 48 c1 e0 04 48 03 47 18 bf 01 00 00 00 48 8b 50 08 48 89 95 38 01 00 00 <48> 89 2a 48 8d 95 30 01 00 00 48 89 50 08 67 e8 53 75 02 00 8b 55
```

journalctl
```
Jan 30 10:35:30 xps7390 systemd-coredump[470461]: Process 441303 (tmux: server) of user 1000 dumped core.

                                                  Stack trace of thread 441303:
                                                  #0  0x000055b11f3f6a79 n/a (tmux + 0x4ca79)
                                                  #1  0x000055b11f3e4cf6 n/a (tmux + 0x3acf6)
                                                  #2  0x000055b11f3e4e00 n/a (tmux + 0x3ae00)
                                                  #3  0x000055b11f41912c n/a (tmux + 0x6f12c)
                                                  #4  0x00007f7dabcce678 bufferevent_run_readcb_ (libevent-2.1.so.7 + 0x16678)
                                                  #5  0x00007f7dabcd3705 n/a (libevent-2.1.so.7 + 0x1b705)
                                                  #6  0x00007f7dabcd9e2a n/a (libevent-2.1.so.7 + 0x21e2a)
                                                  #7  0x00007f7dabcda4cf event_base_loop (libevent-2.1.so.7 + 0x224cf)
                                                  #8  0x000055b11f3f26f0 n/a (tmux + 0x486f0)
                                                  #9  0x000055b11f3fe6fa n/a (tmux + 0x546fa)
                                                  #10 0x000055b11f3bed57 n/a (tmux + 0x14d57)
                                                  #11 0x000055b11f3bb4c6 n/a (tmux + 0x114c6)
                                                  #12 0x00007f7dabaff153 __libc_start_main (libc.so.6 + 0x27153)
                                                  #13 0x000055b11f3bb66e n/a (tmux + 0x1166e)
```

Using sway v.14 (wayland tiling wm) & tmux 3.0a within alacritty v0.4.1, I execute `yay -S chromium-beta-ozone --sudoloop`  in a pane within a small alacritty window, and tmux will always crash while pacman is compiling chromium if toggling the sway windows.
I haven't been able to install chromium yet because it takes hours to compile and I inevitably end up moving an alacritty window by the time it finishes.

**Comment by nicm** [comment]
This is not the same problem, please reproduce from master and open another issue with logs and preferably a backtrace from gdb.

**Comment by nicm** [comment]
I'm going to close this because it doesn't seem to have happened again and without reproducing I don't really know how to fix it.
